File: Scripts/interactive-test.js
#!/usr/bin/env node

import { spawn } from 'child_process';
import readline from 'readline';

class InteractiveMCPTester {
  constructor() {
    this.server = null;
    this.requestId = 1;
    this.rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
  }

  async startServer() {
    console.log('🚀 Starting Salesforce MCP Server...\n');
    
    this.server = spawn('node', ['server.js'], {
      stdio: ['pipe', 'pipe', 'pipe']
    });

    let serverReady = false;
    
    this.server.stderr.on('data', (data) => {
      const output = data.toString();
      if (output.includes('Salesforce MCP server running on stdio')) {
        console.log('✅ Server is running and ready!\n');
        serverReady = true;
        this.showMenu();
      }
    });

    this.server.on('error', (error) => {
      console.error('❌ Server error:', error);
    });

    // Wait for server to start
    await new Promise((resolve) => {
      const checkReady = () => {
        if (serverReady) {
          resolve();
        } else {
          setTimeout(checkReady, 100);
        }
      };
      checkReady();
    });
  }

  showMenu() {
    console.log('🔧 Available Tests:');
    console.log('==================');
    console.log('1. List all available tools');
    console.log('2. Query Salesforce (custom SOQL)');
    console.log('3. Get all Salesforce objects');
    console.log('4. Get recent records');
    console.log('5. Quick Account query');
    console.log('6. Quick Opportunity query');
    console.log('0. Exit');
    console.log();
    
    this.promptUser();
  }

  promptUser() {
    this.rl.question('Choose an option (0-6): ', (answer) => {
      this.handleChoice(answer.trim());
    });
  }

  async handleChoice(choice) {
    console.log();
    
    switch (choice) {
      case '1':
        await this.listTools();
        break;
      case '2':
        await this.customQuery();
        break;
      case '3':
        await this.getSObjects();
        break;
      case '4':
        await this.getRecent();
        break;
      case '5':
        await this.quickAccountQuery();
        break;
      case '6':
        await this.quickOpportunityQuery();
        break;
      case '0':
        this.cleanup();
        return;
      default:
        console.log('❌ Invalid choice. Please try again.\n');
        this.showMenu();
        return;
    }
    
    console.log('\n' + '='.repeat(50) + '\n');
    this.showMenu();
  }

  async sendRequest(method, params = {}) {
    const request = {
      jsonrpc: '2.0',
      id: this.requestId++,
      method,
      params
    };

    return new Promise((resolve, reject) => {
      let responseBuffer = '';
      
      const onData = (data) => {
        responseBuffer += data.toString();
        try {
          const response = JSON.parse(responseBuffer.trim());
          if (response.id === request.id) {
            this.server.stdout.off('data', onData);
            resolve(response);
          }
        } catch (e) {
          // Continue collecting data
        }
      };

      this.server.stdout.on('data', onData);
      
      // Send request
      this.server.stdin.write(JSON.stringify(request) + '\n');
      
      // Timeout after 10 seconds
      setTimeout(() => {
        this.server.stdout.off('data', onData);
        reject(new Error('Request timeout'));
      }, 10000);
    });
  }

  async listTools() {
    console.log('📋 Listing available tools...');
    
    try {
      const response = await this.sendRequest('tools/list');
      
      if (response.result?.tools) {
        console.log(`✅ Found ${response.result.tools.length} tools:`);
        response.result.tools.forEach((tool, index) => {
          console.log(`\n${index + 1}. ${tool.name}`);
          console.log(`   Description: ${tool.description}`);
          if (tool.inputSchema?.properties) {
            console.log(`   Parameters: ${Object.keys(tool.inputSchema.properties).join(', ')}`);
          }
        });
      } else {
        console.log('❌ No tools found');
      }
    } catch (error) {
      console.log('❌ Error:', error.message);
    }
  }

  async customQuery() {
    return new Promise((resolve) => {
      this.rl.question('Enter SOQL query: ', async (query) => {
        console.log(`\n🔍 Executing: ${query}`);
        
        try {
          const response = await this.sendRequest('tools/call', {
            name: 'salesforce_query',
            arguments: { q: query }
          });
          
          if (response.result?.content?.[0]?.text) {
            const result = JSON.parse(response.result.content[0].text);
            console.log('✅ Query Results:');
            console.log(`   Total Size: ${result.totalSize}`);
            console.log(`   Records: ${result.records?.length || 0}`);
            
            if (result.records?.length > 0) {
              console.log('\n   Sample Record:');
              console.log('   ' + JSON.stringify(result.records[0], null, 2).replace(/\n/g, '\n   '));
            }
          } else if (response.result?.content?.[0]?.text?.includes('Error')) {
            console.log('❌ Query Error:', response.result.content[0].text);
          } else {
            console.log('⚠️  Unexpected response format');
          }
        } catch (error) {
          console.log('❌ Error:', error.message);
        }
        
        resolve();
      });
    });
  }

  async getSObjects() {
    console.log('📊 Getting Salesforce objects...');
    
    try {
      const response = await this.sendRequest('tools/call', {
        name: 'salesforce_sobjects',
        arguments: {}
      });
      
      if (response.result?.content?.[0]?.text) {
        const result = JSON.parse(response.result.content[0].text);
        console.log('✅ Salesforce Objects:');
        console.log(`   Total Objects: ${result.sobjects?.length || 0}`);
        
        if (result.sobjects?.length > 0) {
          console.log('\n   Sample Objects:');
          result.sobjects.slice(0, 5).forEach(obj => {
            console.log(`   - ${obj.name} (${obj.label})`);
          });
          
          if (result.sobjects.length > 5) {
            console.log(`   ... and ${result.sobjects.length - 5} more`);
          }
        }
      } else {
        console.log('❌ No objects returned');
      }
    } catch (error) {
      console.log('❌ Error:', error.message);
    }
  }

  async getRecent() {
    console.log('📈 Getting recent records...');
    
    try {
      const response = await this.sendRequest('tools/call', {
        name: 'salesforce_recent',
        arguments: { limit: 10 }
      });
      
      if (response.result?.content?.[0]?.text) {
        const result = JSON.parse(response.result.content[0].text);
        console.log('✅ Recent Records:');
        
        if (result.length > 0) {
          result.forEach((record, index) => {
            console.log(`   ${index + 1}. ${record.Name || record.attributes?.type || 'Unknown'}`);
          });
        } else {
          console.log('   No recent records found');
        }
      } else {
        console.log('❌ No recent records returned');
      }
    } catch (error) {
      console.log('❌ Error:', error.message);
    }
  }

  async quickAccountQuery() {
    console.log('👥 Quick Account Query...');
    
    try {
      const response = await this.sendRequest('tools/call', {
        name: 'salesforce_query',
        arguments: { q: 'SELECT Id, Name, Type, Industry FROM Account LIMIT 5' }
      });
      
      if (response.result?.content?.[0]?.text) {
        const result = JSON.parse(response.result.content[0].text);
        console.log('✅ Account Results:');
        console.log(`   Total Accounts: ${result.totalSize}`);
        
        if (result.records?.length > 0) {
          result.records.forEach((account, index) => {
            console.log(`   ${index + 1}. ${account.Name} (${account.Type || 'No Type'}) - ${account.Industry || 'No Industry'}`);
          });
        }
      } else {
        console.log('❌ No accounts returned');
      }
    } catch (error) {
      console.log('❌ Error:', error.message);
    }
  }

  async quickOpportunityQuery() {
    console.log('💰 Quick Opportunity Query...');
    
    try {
      const response = await this.sendRequest('tools/call', {
        name: 'salesforce_query',
        arguments: { q: 'SELECT Id, Name, StageName, Amount FROM Opportunity LIMIT 5' }
      });
      
      if (response.result?.content?.[0]?.text) {
        const result = JSON.parse(response.result.content[0].text);
        console.log('✅ Opportunity Results:');
        console.log(`   Total Opportunities: ${result.totalSize}`);
        
        if (result.records?.length > 0) {
          result.records.forEach((opp, index) => {
            console.log(`   ${index + 1}. ${opp.Name} - ${opp.StageName} ($${opp.Amount || '0'})`);
          });
        }
      } else {
        console.log('❌ No opportunities returned');
      }
    } catch (error) {
      console.log('❌ Error:', error.message);
    }
  }

  cleanup() {
    console.log('\n🧹 Shutting down...');
    this.rl.close();
    
    if (this.server) {
      this.server.kill('SIGTERM');
    }
    
    console.log('👋 Thanks for testing the Salesforce MCP Connector!');
    process.exit(0);
  }
}

// Start the interactive tester
const tester = new InteractiveMCPTester();
tester.startServer().catch(console.error);

File: Scripts/interactive_test.py
#!/usr/bin/env python3
"""Interactive MCP tester for Salesforce connector using stdio."""

import json
import subprocess
import sys
from pathlib import Path

project_root = Path(__file__).parent.parent
venv_python = project_root / ".venv" / "bin" / "python"

class MCPClient:
    """Simple MCP client using stdio."""
    
    def __init__(self):
        self.process = None
        self.request_id = 0
        self._start_server()
        self._initialize()
    
    def _start_server(self):
        """Start the MCP server in stdio mode."""
        env = {}
        env["MCP_TRANSPORT"] = "stdio"
        
        self.process = subprocess.Popen(
            [str(venv_python), str(project_root / "server.py")],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            cwd=str(project_root),
            env={**dict(subprocess.os.environ), **env}
        )
    
    def _send_jsonrpc(self, method: str, params=None):
        """Send a JSON-RPC 2.0 request and get the response."""
        self.request_id += 1
        
        payload = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
        }
        if params:
            payload["params"] = params
        
        try:
            self.process.stdin.write(json.dumps(payload) + "\n")
            self.process.stdin.flush()
            
            response_line = self.process.stdout.readline()
            if not response_line:
                raise RuntimeError("Server closed connection")
            
            return json.loads(response_line)
        except Exception as e:
            raise RuntimeError(f"MCP communication error: {e}")
    
    def _initialize(self):
        """Initialize the MCP connection."""
        try:
            response = self._send_jsonrpc("initialize", {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {
                    "name": "interactive-test-script",
                    "version": "1.0.0"
                }
            })
            if response.get("error"):
                raise RuntimeError(f"Initialize failed: {response['error']}")
        except Exception as e:
            self.close()
            raise
    
    def list_tools(self):
        """List available tools."""
        response = self._send_jsonrpc("tools/list")
        return response
    
    def call_tool(self, tool_name: str, arguments: dict):
        """Call a tool on the MCP server."""
        response = self._send_jsonrpc("tools/call", {
            "name": tool_name,
            "arguments": arguments
        })
        return response
    
    def close(self):
        """Close the server process."""
        if self.process:
            self.process.terminate()
            try:
                self.process.wait(timeout=2)
            except subprocess.TimeoutExpired:
                self.process.kill()

class InteractiveMCPTester:
    """Interactive test interface for MCP tools."""
    
    def __init__(self):
        self.client = None
    
    def show_menu(self):
        """Display the menu."""
        print("🔧 Available Tests:")
        print("==================")
        print("1. List all available tools")
        print("2. Query Salesforce (custom SOQL)")
        print("3. Get all Salesforce objects")
        print("4. Get recent records")
        print("5. Quick Account query")
        print("6. Quick Opportunity query")
        print("0. Exit")
        print()
    
    def list_tools(self):
        """List all available tools."""
        print("📋 Listing available tools...")
        
        try:
            response = self.client.list_tools()
            
            if response.get("error"):
                print(f"❌ Error: {response['error']}")
                return
            
            tools = response.get("result", {}).get("tools", [])
            print(f"✅ Found {len(tools)} tools:")
            
            for i, tool in enumerate(tools, 1):
                print(f"\n{i}. {tool['name']}")
                print(f"   Description: {tool['description']}")
                if tool.get("inputSchema", {}).get("properties"):
                    params = list(tool["inputSchema"]["properties"].keys())
                    print(f"   Parameters: {', '.join(params)}")
        except Exception as e:
            print(f"❌ Error: {e}")
    
    def custom_query(self):
        """Execute a custom SOQL query."""
        query = input("Enter SOQL query: ").strip()
        
        if not query:
            print("❌ Query cannot be empty")
            return
        
        print(f"\n🔍 Executing: {query}")
        
        try:
            response = self.client.call_tool("salesforce_query", {"q": query})
            
            if response.get("error"):
                print(f"❌ Query Error: {response['error']}")
                return
            
            content = response.get("result", {}).get("content", [{}])[0].get("text", "")
            
            if content:
                try:
                    result = json.loads(content)
                    print("✅ Query Results:")
                    print(f"   Total Size: {result.get('totalSize', 0)}")
                    print(f"   Records: {len(result.get('records', []))}")
                    
                    if result.get("records"):
                        print("\n   Sample Record:")
                        sample = json.dumps(result["records"][0], indent=2)
                        for line in sample.split("\n"):
                            print(f"   {line}")
                except json.JSONDecodeError:
                    print(f"⚠️  Response: {content}")
        except Exception as e:
            print(f"❌ Error: {e}")
    
    def get_sobjects(self):
        """Get all Salesforce objects."""
        print("📊 Getting Salesforce objects...")
        
        try:
            response = self.client.call_tool("salesforce_sobjects", {})
            
            if response.get("error"):
                print(f"❌ Error: {response['error']}")
                return
            
            content = response.get("result", {}).get("content", [{}])[0].get("text", "")
            
            if content:
                try:
                    result = json.loads(content)
                    sobjects = result.get("sobjects", [])
                    print(f"✅ Salesforce Objects:")
                    print(f"   Total Objects: {len(sobjects)}")
                    
                    if sobjects:
                        print("\n   Sample Objects:")
                        for obj in sobjects[:5]:
                            print(f"   - {obj.get('name')} ({obj.get('label')})")
                        
                        if len(sobjects) > 5:
                            print(f"   ... and {len(sobjects) - 5} more")
                except json.JSONDecodeError:
                    print(f"⚠️  Response: {content}")
        except Exception as e:
            print(f"❌ Error: {e}")
    
    def get_recent(self):
        """Get recent records."""
        print("📈 Getting recent records...")
        
        try:
            response = self.client.call_tool("salesforce_recent", {"limit": 10})
            
            if response.get("error"):
                print(f"❌ Error: {response['error']}")
                return
            
            content = response.get("result", {}).get("content", [{}])[0].get("text", "")
            
            if content:
                try:
                    result = json.loads(content)
                    print("✅ Recent Records:")
                    
                    if isinstance(result, list) and len(result) > 0:
                        for i, record in enumerate(result, 1):
                            name = record.get("Name") or record.get("attributes", {}).get("type", "Unknown")
                            print(f"   {i}. {name}")
                    else:
                        print("   No recent records found")
                except json.JSONDecodeError:
                    print(f"⚠️  Response: {content}")
        except Exception as e:
            print(f"❌ Error: {e}")
    
    def quick_account_query(self):
        """Quick account query."""
        print("👥 Quick Account Query...")
        
        try:
            response = self.client.call_tool("salesforce_query", {
                "q": "SELECT Id, Name, Type, Industry FROM Account LIMIT 5"
            })
            
            if response.get("error"):
                print(f"❌ Error: {response['error']}")
                return
            
            content = response.get("result", {}).get("content", [{}])[0].get("text", "")
            
            if content:
                try:
                    result = json.loads(content)
                    print("✅ Account Results:")
                    print(f"   Total Accounts: {result.get('totalSize', 0)}")
                    
                    for i, account in enumerate(result.get("records", []), 1):
                        name = account.get("Name", "Unknown")
                        account_type = account.get("Type", "No Type")
                        industry = account.get("Industry", "No Industry")
                        print(f"   {i}. {name} ({account_type}) - {industry}")
                except json.JSONDecodeError:
                    print(f"⚠️  Response: {content}")
        except Exception as e:
            print(f"❌ Error: {e}")
    
    def quick_opportunity_query(self):
        """Quick opportunity query."""
        print("💰 Quick Opportunity Query...")
        
        try:
            response = self.client.call_tool("salesforce_query", {
                "q": "SELECT Id, Name, StageName, Amount FROM Opportunity LIMIT 5"
            })
            
            if response.get("error"):
                print(f"❌ Error: {response['error']}")
                return
            
            content = response.get("result", {}).get("content", [{}])[0].get("text", "")
            
            if content:
                try:
                    result = json.loads(content)
                    print("✅ Opportunity Results:")
                    print(f"   Total Opportunities: {result.get('totalSize', 0)}")
                    
                    for i, opp in enumerate(result.get("records", []), 1):
                        name = opp.get("Name", "Unknown")
                        stage = opp.get("StageName", "Unknown")
                        amount = opp.get("Amount", "0")
                        print(f"   {i}. {name} - {stage} (${amount})")
                except json.JSONDecodeError:
                    print(f"⚠️  Response: {content}")
        except Exception as e:
            print(f"❌ Error: {e}")
    
    def handle_choice(self, choice: str) -> bool:
        """Handle user menu choice. Returns False if user wants to exit."""
        print()
        
        if choice == "1":
            self.list_tools()
        elif choice == "2":
            self.custom_query()
        elif choice == "3":
            self.get_sobjects()
        elif choice == "4":
            self.get_recent()
        elif choice == "5":
            self.quick_account_query()
        elif choice == "6":
            self.quick_opportunity_query()
        elif choice == "0":
            return False
        else:
            print("❌ Invalid choice. Please try again.\n")
            return True
        
        return True
    
    def run(self):
        """Run the interactive tester."""
        print("🚀 Starting Salesforce MCP Server...\n")
        
        try:
            self.client = MCPClient()
            print("✅ Connected to MCP server!\n")
            
            while True:
                self.show_menu()
                choice = input("Choose an option (0-6): ").strip()
                
                should_continue = self.handle_choice(choice)
                if not should_continue:
                    break
                
                print("\n" + "=" * 50 + "\n")
        
        except Exception as e:
            print(f"❌ Error: {e}")
        finally:
            if self.client:
                self.client.close()
            
            print("\n🧹 Shutting down...")
            print("👋 Thanks for testing the Salesforce MCP Connector!")

def main():
    """Main function."""
    tester = InteractiveMCPTester()
    tester.run()

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\n👋 Exiting...")
        sys.exit(0)
